File: apps/mobile-app/ios/AliasVaultUITests/TestUserRegistration.swift
import Foundation
import CryptoKit
import SQLite3

// Import the Rust core SRP functions
import RustCoreFramework

// Import VaultModels for VaultVersions and VaultSql
import VaultModels

/// Test user credentials and token.
struct TestUser {
    let username: String
    let password: String
    let token: String
    let refreshToken: String
}

/// API response models
struct TokenResponse: Codable {
    let token: String
    let refreshToken: String
}

struct ApiError: Codable {
    let title: String?
    let message: String?
}

/// Default encryption settings for Argon2Id.
/// These match the server defaults in AliasVault.Cryptography.Client/Defaults.cs
enum EncryptionDefaults {
    static let type = "Argon2Id"
    static let iterations: UInt32 = 2
    static let memorySize: UInt32 = 19456
    static let parallelism: UInt32 = 1

    static var settingsJson: String {
        """
        {"DegreeOfParallelism":\(parallelism),"MemorySize":\(memorySize),"Iterations":\(iterations)}
        """
    }
}

/// Test user registration helper using SRP protocol.
/// Mirrors the browser extension's test-api.ts implementation.
enum TestUserRegistration {

    /// API URL for testing (defaults to local development server)
    static var apiUrl: String {
        return ProcessInfo.processInfo.environment["API_URL"] ?? "http://localhost:5092"
    }

    // MARK: - Username/Password Generation

    /// Generate a random test username.
    static func generateTestUsername() -> String {
        let randomPart = String((0..<10).map { _ in "abcdefghijklmnopqrstuvwxyz0123456789".randomElement()! })
        return "test_\(randomPart)@example.tld"
    }

    /// Generate a test password.
    /// Uses a static password for easier debugging and test reproducibility.
    static func generateTestPassword() -> String {
        return "password"
    }

    /// Normalize username by converting to lowercase and trimming whitespace.
    static func normalizeUsername(_ username: String) -> String {
        return username.lowercased().trimmingCharacters(in: .whitespaces)
    }

    // MARK: - Argon2 Key Derivation

    /// Derive encryption key from password using Argon2Id via Rust core.
    ///
    /// Uses the AliasVault default parameters:
    /// - Iterations: 2
    /// - Memory: 19456 KiB
    /// - Parallelism: 1
    /// - Output length: 32 bytes
    static func deriveKeyArgon2(_ password: String, salt: String) throws -> Data {
        // Use the Rust core's Argon2 implementation
        let hashHex = try argon2HashPassword(password: password, salt: salt)

        // Convert hex string to Data
        return try hexToBytes(hashHex)
    }

    // MARK: - Hex Conversion

    /// Convert Data to uppercase hex string.
    static func bytesToHex(_ data: Data) -> String {
        return data.map { String(format: "%02X", $0) }.joined()
    }

    /// Convert hex string to Data.
    static func hexToBytes(_ hex: String) throws -> Data {
        var hex = hex.trimmingCharacters(in: .whitespaces)
        if hex.hasPrefix("0x") || hex.hasPrefix("0X") {
            hex = String(hex.dropFirst(2))
        }

        guard hex.count % 2 == 0 else {
            throw NSError(domain: "TestUserRegistration", code: 3,
                         userInfo: [NSLocalizedDescriptionKey: "Invalid hex string length"])
        }

        var data = Data()
        var index = hex.startIndex
        while index < hex.endIndex {
            let nextIndex = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<nextIndex], radix: 16) else {
                throw NSError(domain: "TestUserRegistration", code: 4,
                             userInfo: [NSLocalizedDescriptionKey: "Invalid hex character"])
            }
            data.append(byte)
            index = nextIndex
        }
        return data
    }

    // MARK: - Registration

    /// Register a new test user via the API using SRP protocol.
    static func registerTestUser(
        apiBaseUrl: String,
        username: String,
        password: String
    ) async throws -> TokenResponse {
        let baseUrl = apiBaseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/v1/"
        let normalizedUsername = normalizeUsername(username)

        // Generate salt using Rust core
        let salt = srpGenerateSalt()

        // Derive key from password
        let encryptionKey = try deriveKeyArgon2(password, salt: salt)
        let passwordHashHex = bytesToHex(encryptionKey)

        // Derive SRP private key and verifier using Rust core
        let privateKey = try srpDerivePrivateKey(salt: salt, identity: normalizedUsername, passwordHash: passwordHashHex)
        let verifier = try srpDeriveVerifier(privateKey: privateKey)

        // Build registration request
        let registerRequest: [String: Any] = [
            "username": normalizedUsername,
            "salt": salt,
            "verifier": verifier,
            "encryptionType": EncryptionDefaults.type,
            "encryptionSettings": EncryptionDefaults.settingsJson
        ]

        let requestData = try JSONSerialization.data(withJSONObject: registerRequest)

        // Send registration request
        var request = URLRequest(url: URL(string: "\(baseUrl)Auth/register")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = requestData

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "TestUserRegistration", code: 5,
                         userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        guard httpResponse.statusCode == 200 else {
            let errorMessage: String
            if let apiError = try? JSONDecoder().decode(ApiError.self, from: data) {
                errorMessage = apiError.title ?? apiError.message ?? "Unknown error"
            } else if let text = String(data: data, encoding: .utf8) {
                errorMessage = text
            } else {
                errorMessage = "HTTP \(httpResponse.statusCode)"
            }
            throw NSError(domain: "TestUserRegistration", code: httpResponse.statusCode,
                         userInfo: [NSLocalizedDescriptionKey: "Registration failed: \(errorMessage)"])
        }

        let tokenResponse = try JSONDecoder().decode(TokenResponse.self, from: data)

        // Upload initial empty vault
        try await uploadInitialVault(
            apiBaseUrl: apiBaseUrl,
            token: tokenResponse.token,
            username: normalizedUsername,
            encryptionKey: encryptionKey
        )

        return tokenResponse
    }

    // MARK: - Vault Upload

    /// Upload an initial empty vault to the server.
    static func uploadInitialVault(
        apiBaseUrl: String,
        token: String,
        username: String,
        encryptionKey: Data
    ) async throws {
        let baseUrl = apiBaseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/v1/"

        // Create empty vault database (minimal SQLite)
        let vaultBase64 = createEmptyVaultDatabase()

        // Encrypt the vault using AES-GCM
        let encryptedVault = try symmetricEncrypt(plaintext: vaultBase64, key: encryptionKey)

        // Generate RSA key pair for the vault
        let rsaKeyPair = try generateRsaKeyPair()

        // Get current timestamp
        let now = ISO8601DateFormatter().string(from: Date())

        // Build vault upload request
        let vaultRequest: [String: Any] = [
            "username": normalizeUsername(username),
            "blob": encryptedVault,
            "version": VaultVersions.latestVersion, // From auto-generated VaultVersions.swift
            "currentRevisionNumber": VaultVersions.latestRevision,
            "encryptionPublicKey": rsaKeyPair.publicKey,
            "credentialsCount": 0,
            "emailAddressList": [String](),
            "privateEmailDomainList": [String](),
            "hiddenPrivateEmailDomainList": [String](),
            "publicEmailDomainList": [String](),
            "createdAt": now,
            "updatedAt": now
        ]

        let requestData = try JSONSerialization.data(withJSONObject: vaultRequest)

        var request = URLRequest(url: URL(string: "\(baseUrl)Vault")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = requestData

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            let errorText = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw NSError(domain: "TestUserRegistration", code: 6,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to upload vault: \(errorText)"])
        }
    }

    // MARK: - Encryption Helpers

    /// Encrypt data using AES-GCM (matches browser extension's EncryptionUtility).
    static func symmetricEncrypt(plaintext: String, key: Data) throws -> String {
        guard let plaintextData = plaintext.data(using: .utf8) else {
            throw NSError(domain: "TestUserRegistration", code: 7,
                         userInfo: [NSLocalizedDescriptionKey: "Invalid plaintext encoding"])
        }

        let symmetricKey = SymmetricKey(data: key)
        let sealedBox = try AES.GCM.seal(plaintextData, using: symmetricKey)

        // Combined format: nonce + ciphertext + tag
        guard let combined = sealedBox.combined else {
            throw NSError(domain: "TestUserRegistration", code: 8,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to get combined sealed box"])
        }

        return combined.base64EncodedString()
    }

    /// Generate RSA key pair for vault encryption.
    static func generateRsaKeyPair() throws -> (publicKey: String, privateKey: String) {
        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeySizeInBits as String: 2048
        ]

        var error: Unmanaged<CFError>?
        guard let privateKey = SecKeyCreateRandomKey(attributes as CFDictionary, &error) else {
            throw error!.takeRetainedValue() as Error
        }

        guard let publicKey = SecKeyCopyPublicKey(privateKey) else {
            throw NSError(domain: "TestUserRegistration", code: 9,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to get public key"])
        }

        // Export keys as JWK format
        let publicKeyJwk = try exportKeyAsJwk(publicKey, isPrivate: false)
        let privateKeyJwk = try exportKeyAsJwk(privateKey, isPrivate: true)

        return (publicKeyJwk, privateKeyJwk)
    }

    /// Export a SecKey as JWK JSON string.
    static func exportKeyAsJwk(_ key: SecKey, isPrivate: Bool) throws -> String {
        var error: Unmanaged<CFError>?
        guard let keyData = SecKeyCopyExternalRepresentation(key, &error) as Data? else {
            throw error!.takeRetainedValue() as Error
        }

        // For RSA keys, create a simplified JWK
        // This is a basic implementation - production would need full JWK encoding
        let base64Key = keyData.base64EncodedString()

        let jwk: [String: Any] = [
            "kty": "RSA",
            "key_ops": isPrivate ? ["decrypt"] : ["encrypt"],
            "ext": true,
            "key": base64Key
        ]

        let jsonData = try JSONSerialization.data(withJSONObject: jwk)
        return String(data: jsonData, encoding: .utf8)!
    }

    // MARK: - Empty Vault Creation

    /// Create an empty vault database as base64 string.
    /// This creates a real SQLite database with the vault schema using VaultSql.completeSchema.
    static func createEmptyVaultDatabase() -> String {
        // Create a temporary SQLite database file
        let tempDir = FileManager.default.temporaryDirectory
        let dbPath = tempDir.appendingPathComponent(UUID().uuidString + ".db").path

        var database: OpaquePointer?

        // Open/create the database
        guard sqlite3_open(dbPath, &database) == SQLITE_OK else {
            print("Failed to create SQLite database")
            return ""
        }

        defer {
            sqlite3_close(database)
            try? FileManager.default.removeItem(atPath: dbPath)
        }

        // Execute the complete schema SQL from VaultSql (auto-generated from TypeScript)
        let schemaSql = VaultSql.completeSchema
        var errorMessage: UnsafeMutablePointer<CChar>?

        let result = sqlite3_exec(database, schemaSql, nil, nil, &errorMessage)
        if result != SQLITE_OK {
            let error = errorMessage.map { String(cString: $0) } ?? "Unknown error"
            print("Failed to execute schema SQL: \(error)")
            sqlite3_free(errorMessage)
            return ""
        }

        // Read the database file and encode as base64
        guard let dbData = FileManager.default.contents(atPath: dbPath) else {
            print("Failed to read database file")
            return ""
        }

        return dbData.base64EncodedString()
    }

    // MARK: - Public API

    /// Create a test user with random credentials.
    static func createTestUser(apiBaseUrl: String? = nil) async throws -> TestUser {
        let url = apiBaseUrl ?? apiUrl
        let username = generateTestUsername()
        let password = generateTestPassword()

        let tokenResponse = try await registerTestUser(
            apiBaseUrl: url,
            username: username,
            password: password
        )

        return TestUser(
            username: username,
            password: password,
            token: tokenResponse.token,
            refreshToken: tokenResponse.refreshToken
        )
    }

    /// Check if the API is available.
    static func isApiAvailable(apiBaseUrl: String? = nil) async -> Bool {
        let url = (apiBaseUrl ?? apiUrl).trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/v1/"

        var request = URLRequest(url: URL(string: "\(url)Auth/status")!)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 5

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            if let httpResponse = response as? HTTPURLResponse {
                // Status endpoint returns 401 when not authenticated, but that means API is running
                return httpResponse.statusCode == 401 || httpResponse.statusCode == 200
            }
            return false
        } catch {
            return false
        }
    }

    // MARK: - Test Helpers (DEV API Endpoints)

    /// Delete the newest vault revisions for the authenticated user.
    /// This endpoint only works in development mode.
    /// Used for testing RPO (Recovery Point Objective) recovery scenarios.
    ///
    /// - Parameters:
    ///   - count: Number of newest revisions to delete
    ///   - token: Authentication token
    ///   - apiBaseUrl: Optional API base URL (defaults to apiUrl)
    /// - Returns: Number of deleted revisions
    static func deleteVaultRevisions(
        count: Int,
        token: String,
        apiBaseUrl: String? = nil
    ) async throws -> Int {
        let url = (apiBaseUrl ?? apiUrl).trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/v1/"

        var request = URLRequest(url: URL(string: "\(url)Test/vault-revisions/\(count)")!)
        request.httpMethod = "DELETE"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "TestUserRegistration", code: 10,
                         userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        guard httpResponse.statusCode == 200 else {
            let errorText = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw NSError(domain: "TestUserRegistration", code: httpResponse.statusCode,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to delete vault revisions: \(errorText)"])
        }

        // Parse response to get deleted count
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let deleted = json["deleted"] as? Int {
            return deleted
        }

        return 0
    }

    /// Get vault revision information for the authenticated user.
    /// This endpoint only works in development mode.
    ///
    /// - Parameters:
    ///   - token: Authentication token
    ///   - apiBaseUrl: Optional API base URL (defaults to apiUrl)
    /// - Returns: Tuple of (count, currentRevision)
    static func getVaultRevisions(
        token: String,
        apiBaseUrl: String? = nil
    ) async throws -> (count: Int, currentRevision: Int) {
        let url = (apiBaseUrl ?? apiUrl).trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/v1/"

        var request = URLRequest(url: URL(string: "\(url)Test/vault-revisions")!)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "TestUserRegistration", code: 10,
                         userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        guard httpResponse.statusCode == 200 else {
            let errorText = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw NSError(domain: "TestUserRegistration", code: httpResponse.statusCode,
                         userInfo: [NSLocalizedDescriptionKey: "Failed to get vault revisions: \(errorText)"])
        }

        // Parse response
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let count = json["count"] as? Int,
           let currentRevision = json["currentRevision"] as? Int {
            return (count, currentRevision)
        }

        return (0, 0)
    }
}

File: apps/mobile-app/android/app/src/androidTest/java/net/aliasvault/app/TestUserRegistration.kt
package net.aliasvault.app

import android.database.sqlite.SQLiteDatabase
import android.util.Base64
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.aliasvault.app.vaultstore.models.VaultSql
import net.aliasvault.app.vaultstore.models.VaultVersions
import org.json.JSONArray
import org.json.JSONObject
import uniffi.aliasvault_core.argon2HashPassword
import uniffi.aliasvault_core.srpDerivePrivateKey
import uniffi.aliasvault_core.srpDeriveVerifier
import uniffi.aliasvault_core.srpGenerateSalt
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.security.KeyPairGenerator
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Test user credentials and token.
 */
data class TestUser(
    val username: String,
    val password: String,
    val token: String,
    val refreshToken: String,
)

/**
 * Token response from the API.
 */
data class TokenResponse(
    val token: String,
    val refreshToken: String,
)

/**
 * Test user registration helper using SRP protocol.
 * Uses the Rust core library for SRP operations via UniFFI bindings.
 */
object TestUserRegistration {
    private const val TAG = "TestUserRegistration"

    /**
     * API URL for testing.
     */
    val apiUrl: String
        get() = TestConfiguration.apiUrl

    // region Username/Password Generation

    /**
     * Generate a random test username.
     */
    fun generateTestUsername(): String {
        val chars = "abcdefghijklmnopqrstuvwxyz0123456789"
        val randomPart = (1..10)
            .map { chars.random() }
            .joinToString("")
        return "test_$randomPart@example.tld"
    }

    /**
     * Generate a test password.
     * Uses a static password for easier debugging and test reproducibility.
     */
    fun generateTestPassword(): String = "password"

    /**
     * Normalize username by converting to lowercase and trimming whitespace.
     */
    fun normalizeUsername(username: String): String =
        username.lowercase().trim()

    // endregion

    // region Hex Conversion

    /**
     * Convert ByteArray to uppercase hex string.
     */
    fun bytesToHex(data: ByteArray): String =
        data.joinToString("") { "%02X".format(it) }

    /**
     * Convert hex string to ByteArray.
     */
    fun hexToBytes(hex: String): ByteArray {
        val cleanHex = hex.removePrefix("0x").removePrefix("0X").trim()
        require(cleanHex.length % 2 == 0) { "Invalid hex string length" }

        return ByteArray(cleanHex.length / 2) { i ->
            cleanHex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
    }

    // endregion

    // region Registration

    /**
     * Register a new test user via the API using SRP protocol.
     * Uses Rust core for all SRP operations.
     */
    suspend fun registerTestUser(
        apiBaseUrl: String,
        username: String,
        password: String,
    ): TokenResponse = withContext(Dispatchers.IO) {
        val baseUrl = apiBaseUrl.trimEnd('/') + "/v1/"
        val normalizedUsername = normalizeUsername(username)

        // Generate salt using Rust core
        val salt = srpGenerateSalt()

        // Derive key from password using Rust core Argon2
        val passwordHashHex = argon2HashPassword(password, salt)

        // Derive SRP private key and verifier using Rust core
        val privateKey = srpDerivePrivateKey(salt, normalizedUsername, passwordHashHex)
        val verifier = srpDeriveVerifier(privateKey)

        // Build registration request
        val registerRequest = JSONObject().apply {
            put("username", normalizedUsername)
            put("salt", salt)
            put("verifier", verifier)
            put("encryptionType", TestConfiguration.EncryptionDefaults.TYPE)
            put("encryptionSettings", TestConfiguration.EncryptionDefaults.settingsJson)
        }

        // Send registration request
        val connection = URL("${baseUrl}Auth/register").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.doOutput = true

            connection.outputStream.bufferedWriter().use {
                it.write(registerRequest.toString())
            }

            if (connection.responseCode != 200) {
                val errorBody = try {
                    connection.errorStream?.bufferedReader()?.readText() ?: "Unknown error"
                } catch (e: Exception) {
                    "HTTP ${connection.responseCode}"
                }
                throw Exception("Registration failed: $errorBody")
            }

            val responseBody = connection.inputStream.bufferedReader().readText()
            val json = JSONObject(responseBody)

            val tokenResponse = TokenResponse(
                token = json.getString("token"),
                refreshToken = json.getString("refreshToken"),
            )

            // Upload initial empty vault
            val encryptionKey = hexToBytes(passwordHashHex)
            uploadInitialVault(
                apiBaseUrl = apiBaseUrl,
                token = tokenResponse.token,
                username = normalizedUsername,
                encryptionKey = encryptionKey,
            )

            tokenResponse
        } finally {
            connection.disconnect()
        }
    }

    // endregion

    // region Vault Upload

    /**
     * Upload an initial empty vault to the server.
     */
    private suspend fun uploadInitialVault(
        apiBaseUrl: String,
        token: String,
        username: String,
        encryptionKey: ByteArray,
    ) = withContext(Dispatchers.IO) {
        val baseUrl = apiBaseUrl.trimEnd('/') + "/v1/"

        // Create empty vault database
        val vaultBase64 = createEmptyVaultDatabase()

        // Encrypt the vault using AES-GCM
        val encryptedVault = symmetricEncrypt(vaultBase64, encryptionKey)

        // Generate RSA key pair for the vault
        val rsaKeyPair = generateRsaKeyPair()

        // Get current timestamp in ISO8601 format
        val now = java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", java.util.Locale.US)
            .apply { timeZone = java.util.TimeZone.getTimeZone("UTC") }
            .format(java.util.Date())

        // Build vault upload request
        val vaultRequest = JSONObject().apply {
            put("username", normalizeUsername(username))
            put("blob", encryptedVault)
            put("version", VaultVersions.latestVersion)
            put("currentRevisionNumber", VaultVersions.latestRevision)
            put("encryptionPublicKey", rsaKeyPair.first)
            put("credentialsCount", 0)
            put("emailAddressList", JSONArray())
            put("privateEmailDomainList", JSONArray())
            put("hiddenPrivateEmailDomainList", JSONArray())
            put("publicEmailDomainList", JSONArray())
            put("createdAt", now)
            put("updatedAt", now)
        }

        val connection = URL("${baseUrl}Vault").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.setRequestProperty("Authorization", "Bearer $token")
            connection.doOutput = true

            connection.outputStream.bufferedWriter().use {
                it.write(vaultRequest.toString())
            }

            if (connection.responseCode != 200) {
                val errorBody = try {
                    connection.errorStream?.bufferedReader()?.readText() ?: "Unknown error"
                } catch (e: Exception) {
                    "HTTP ${connection.responseCode}"
                }
                throw Exception("Failed to upload vault: $errorBody")
            }
        } finally {
            connection.disconnect()
        }
    }

    // endregion

    // region Encryption Helpers

    /**
     * Encrypt data using AES-GCM.
     */
    private fun symmetricEncrypt(plaintext: String, key: ByteArray): String {
        val iv = ByteArray(12)
        SecureRandom().nextBytes(iv)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val keySpec = SecretKeySpec(key, "AES")
        val gcmSpec = GCMParameterSpec(128, iv)

        cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec)
        val encrypted = cipher.doFinal(plaintext.toByteArray(StandardCharsets.UTF_8))

        // Combined format: nonce + ciphertext + tag
        val combined = ByteArray(iv.size + encrypted.size)
        System.arraycopy(iv, 0, combined, 0, iv.size)
        System.arraycopy(encrypted, 0, combined, iv.size, encrypted.size)

        return Base64.encodeToString(combined, Base64.NO_WRAP)
    }

    /**
     * Generate RSA key pair for vault encryption.
     */
    private fun generateRsaKeyPair(): Pair<String, String> {
        val keyPairGenerator = KeyPairGenerator.getInstance("RSA")
        keyPairGenerator.initialize(2048)
        val keyPair = keyPairGenerator.generateKeyPair()

        val publicKey = keyPair.public as java.security.interfaces.RSAPublicKey
        val privateKey = keyPair.private as java.security.interfaces.RSAPrivateKey

        // Export as simple JWK format
        val publicKeyJwk = JSONObject().apply {
            put("kty", "RSA")
            put("key_ops", JSONArray().put("encrypt"))
            put("ext", true)
            put(
                "n",
                Base64.encodeToString(
                    publicKey.modulus.toByteArray(),
                    Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP,
                ),
            )
            put(
                "e",
                Base64.encodeToString(
                    publicKey.publicExponent.toByteArray(),
                    Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP,
                ),
            )
        }

        val privateKeyJwk = JSONObject().apply {
            put("kty", "RSA")
            put("key_ops", JSONArray().put("decrypt"))
            put("ext", true)
            put(
                "n",
                Base64.encodeToString(
                    privateKey.modulus.toByteArray(),
                    Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP,
                ),
            )
        }

        return Pair(publicKeyJwk.toString(), privateKeyJwk.toString())
    }

    // endregion

    // region Empty Vault Creation

    /**
     * Create an empty vault database as base64 string.
     */
    private fun createEmptyVaultDatabase(): String {
        val tempFile = File.createTempFile("vault", ".db")
        try {
            val db = SQLiteDatabase.openOrCreateDatabase(tempFile, null)
            try {
                // Execute the complete schema SQL
                val statements = VaultSql.completeSchema.split(";")
                for (statement in statements) {
                    val trimmed = statement.trim()
                    if (trimmed.isNotEmpty()) {
                        db.execSQL("$trimmed;")
                    }
                }
            } finally {
                db.close()
            }

            // Read the database file and encode as base64
            val dbBytes = tempFile.readBytes()
            return Base64.encodeToString(dbBytes, Base64.NO_WRAP)
        } finally {
            tempFile.delete()
        }
    }

    // endregion

    // region Public API

    /**
     * Create a test user with random credentials.
     */
    suspend fun createTestUser(apiBaseUrl: String? = null): TestUser {
        val url = apiBaseUrl ?: apiUrl
        val username = generateTestUsername()
        val password = generateTestPassword()

        val tokenResponse = registerTestUser(
            apiBaseUrl = url,
            username = username,
            password = password,
        )

        return TestUser(
            username = username,
            password = password,
            token = tokenResponse.token,
            refreshToken = tokenResponse.refreshToken,
        )
    }

    /**
     * Check if the API is available.
     */
    suspend fun isApiAvailable(apiBaseUrl: String? = null): Boolean = withContext(Dispatchers.IO) {
        val url = (apiBaseUrl ?: apiUrl).trimEnd('/') + "/v1/"

        try {
            val connection = URL("${url}Auth/status").openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.connectTimeout = 5000
            connection.readTimeout = 5000

            val responseCode = connection.responseCode
            connection.disconnect()

            // Status endpoint returns 401 when not authenticated, but that means API is running
            responseCode == 401 || responseCode == 200
        } catch (e: Exception) {
            Log.e(TAG, "API not available", e)
            false
        }
    }

    // endregion

    // region Test Helpers (DEV API Endpoints)

    /**
     * Get vault revision information for a user by username.
     * This is an anonymous endpoint that doesn't require authentication.
     */
    suspend fun getVaultRevisionsByUsername(
        username: String,
        apiBaseUrl: String? = null,
    ): Pair<Int, Int> = withContext(Dispatchers.IO) {
        val url = (apiBaseUrl ?: apiUrl).trimEnd('/') + "/v1/"
        val encodedUsername = java.net.URLEncoder.encode(username, "UTF-8")

        val connection =
            URL("${url}Test/vault-revisions/by-username/$encodedUsername").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "GET"
            connection.setRequestProperty("Content-Type", "application/json")

            if (connection.responseCode != 200) {
                throw Exception("Failed to get vault revisions: HTTP ${connection.responseCode}")
            }

            val responseBody = connection.inputStream.bufferedReader().readText()
            val json = JSONObject(responseBody)

            Pair(
                json.optInt("count", 0),
                json.optInt("currentRevision", 0),
            )
        } finally {
            connection.disconnect()
        }
    }

    /**
     * Delete the newest vault revisions for a user by username.
     */
    suspend fun deleteVaultRevisionsByUsername(
        username: String,
        count: Int,
        apiBaseUrl: String? = null,
    ): Int = withContext(Dispatchers.IO) {
        val url = (apiBaseUrl ?: apiUrl).trimEnd('/') + "/v1/"
        val encodedUsername = java.net.URLEncoder.encode(username, "UTF-8")

        val connection =
            URL("${url}Test/vault-revisions/by-username/$encodedUsername/$count").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "DELETE"
            connection.setRequestProperty("Content-Type", "application/json")

            if (connection.responseCode != 200) {
                throw Exception("Failed to delete vault revisions: HTTP ${connection.responseCode}")
            }

            val responseBody = connection.inputStream.bufferedReader().readText()
            val json = JSONObject(responseBody)

            json.optInt("deleted", 0)
        } finally {
            connection.disconnect()
        }
    }

    /**
     * Block a user's account by username.
     */
    suspend fun blockUserByUsername(
        username: String,
        apiBaseUrl: String? = null,
    ) = withContext(Dispatchers.IO) {
        val url = (apiBaseUrl ?: apiUrl).trimEnd('/') + "/v1/"
        val encodedUsername = java.net.URLEncoder.encode(username, "UTF-8")

        val connection =
            URL("${url}Test/block-user/by-username/$encodedUsername").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")

            if (connection.responseCode != 200) {
                throw Exception("Failed to block user: HTTP ${connection.responseCode}")
            }
        } finally {
            connection.disconnect()
        }
    }

    /**
     * Unblock a user's account by username.
     */
    suspend fun unblockUserByUsername(
        username: String,
        apiBaseUrl: String? = null,
    ) = withContext(Dispatchers.IO) {
        val url = (apiBaseUrl ?: apiUrl).trimEnd('/') + "/v1/"
        val encodedUsername = java.net.URLEncoder.encode(username, "UTF-8")

        val connection =
            URL("${url}Test/unblock-user/by-username/$encodedUsername").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")

            if (connection.responseCode != 200) {
                throw Exception("Failed to unblock user: HTTP ${connection.responseCode}")
            }
        } finally {
            connection.disconnect()
        }
    }

    // endregion
}
